File: engine/src/main/java/com/terralite/engine/chunk/InMemoryChunkStorage.java
package com.terralite.engine.chunk;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

public final class InMemoryChunkStorage implements ChunkStorage {
    private final Map<ChunkPos, Chunk> chunks = new LinkedHashMap<>();

    @Override
    public Optional<Chunk> get(ChunkPos pos) {
        return Optional.ofNullable(chunks.get(Objects.requireNonNull(pos, "pos")));
    }

    @Override
    public boolean contains(ChunkPos pos) {
        return chunks.containsKey(Objects.requireNonNull(pos, "pos"));
    }

    @Override
    public Chunk put(Chunk chunk) {
        Objects.requireNonNull(chunk, "chunk");
        chunks.put(chunk.pos(), chunk);
        return chunk;
    }

    @Override
    public Chunk remove(ChunkPos pos) {
        Chunk removed = chunks.remove(Objects.requireNonNull(pos, "pos"));
        if (removed == null) {
            throw new IllegalArgumentException("Missing chunk: " + pos);
        }
        return removed;
    }

    @Override
    public Chunk require(ChunkPos pos) {
        return get(pos).orElseThrow(() -> new IllegalArgumentException("Missing chunk: " + pos));
    }

    @Override
    public Collection<ChunkPos> positions() {
        return List.copyOf(chunks.keySet());
    }

    @Override
    public Collection<Chunk> chunks() {
        return List.copyOf(chunks.values());
    }

    @Override
    public int size() {
        return chunks.size();
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/chunk/InMemoryChunkStorage.kt
package com.terralite.engine.chunk

class InMemoryChunkStorage : ChunkStorage {
    private val chunks: MutableMap<ChunkPos, Chunk> = LinkedHashMap()

    override fun get(pos: ChunkPos): Chunk? = chunks[pos]

    override fun contains(pos: ChunkPos): Boolean = chunks.containsKey(pos)

    override fun put(chunk: Chunk): Chunk {
        chunks[chunk.pos] = chunk
        return chunk
    }

    override fun remove(pos: ChunkPos): Chunk {
        return chunks.remove(pos) ?: throw IllegalArgumentException("Missing chunk: $pos")
    }

    override fun require(pos: ChunkPos): Chunk {
        return get(pos) ?: throw IllegalArgumentException("Missing chunk: $pos")
    }

    override fun positions(): Collection<ChunkPos> = chunks.keys.toList()

    override fun chunks(): Collection<Chunk> = chunks.values.toList()

    override fun size(): Int = chunks.size
}
